File: rust/src/lazy_merger.rs
use crate::index_builder::BATCH_SIZE;
use crate::postings::Postings;
use std::fs;
use std::io::{BufRead, BufReader, LineWriter, Write};
use std::path::PathBuf;

use crate::file_skip_list::{FileSkip, MERGED_INDEX_DIR};
pub fn main() {
    fs::create_dir_all(MERGED_INDEX_DIR).unwrap_or_default();
    let doc_id = 55393;
    let batch_count = (doc_id - 1) / BATCH_SIZE + 1;
    let word_ranges = ["0_9", "a_f", "g_p", "q_z"];
    // first character is null
    let mut current_first_char = '\0';
    let mut final_file_appender: Option<LineWriter<fs::File>> = None;

    for &words in &word_ranges {
        // Open all available files for this word range
        let mut readers = Vec::with_capacity(batch_count as usize);

        // Open all available files for this word range
        for i in 0..batch_count {
            let filepath = format!("inverted_index/{}/{}.txt", i, words);
            if let Ok(file) = fs::File::open(&filepath) {
                readers.push((BufReader::new(file), String::new()));
            }
        }

        if readers.is_empty() {
            continue;
        }

        // initialize every single posting with a line from the respective files, if available else remove it
        let mut postings_with_indices = Vec::with_capacity(readers.len());
        for (i, (reader, _)) in readers.iter_mut().enumerate() {
            let mut line = String::new();
            if reader.read_line(&mut line).unwrap_or(0) > 0 && !line.trim().is_empty() {
                if let Ok(posting) = Postings::load_postings(&line) {
                    postings_with_indices.push((posting, i));
                }
            }
        }

        while !postings_with_indices.is_empty() {
            // find smallest word
            postings_with_indices.sort_unstable_by(|a, b| a.0.word.cmp(&b.0.word));
            let smallest_word = postings_with_indices[0].0.word.clone();

            // first character check - create new file if needed
            let first_char = smallest_word.chars().next().unwrap_or('0');
            if first_char != current_first_char {
                if let Some(writer) = final_file_appender.as_mut() {
                    writer.flush().unwrap();
                }
                // Build and write skip list for previous character unless its starting character of null
                if current_first_char != '\0' {
                    // build a skip list on that file
                    let skip_list_path =
                        PathBuf::from(format!("{}/{}.txt", MERGED_INDEX_DIR, current_first_char));
                    let file_skip_list = FileSkip::build_skip_list(skip_list_path);
                    FileSkip::write_skip_list(&file_skip_list);
                }
                // append the postings to the new file
                current_first_char = first_char;
                let file_path = format!("{}/{}.txt", MERGED_INDEX_DIR, current_first_char);
                final_file_appender = Some(LineWriter::new(fs::File::create(file_path).unwrap()));
            }

            // merge all postings with the smallest word
            let mut merged_posting = Postings::new(smallest_word);
            let mut indices_to_update = Vec::new();

            let mut i = 0;
            // merge all postings with the smallest word
            while i < postings_with_indices.len() {
                if postings_with_indices[i].0.word == merged_posting.word {
                    merged_posting.merge(postings_with_indices[i].0.clone());
                    indices_to_update.push(postings_with_indices[i].1);
                    postings_with_indices.remove(i);
                } else {
                    i += 1;
                }
            }

            // Write merged posting
            if let Some(writer) = final_file_appender.as_mut() {
                writer
                    .write_all((merged_posting.save_postings() + "\n").as_bytes())
                    .unwrap();
            }

            // Read next lines for updated readers
            for &idx in &indices_to_update {
                let mut line = String::new();
                if readers[idx].0.read_line(&mut line).unwrap_or(0) > 0 && !line.trim().is_empty() {
                    if let Ok(posting) = Postings::load_postings(&line) {
                        postings_with_indices.push((posting, idx));
                    }
                }
            }
        }
    }

    // Flush and create skip list for the last file
    if let Some(writer) = final_file_appender.as_mut() {
        writer.flush().unwrap();
    }

    if current_first_char != '\0' {
        let skip_list_path =
            PathBuf::from(format!("{}/{}.txt", MERGED_INDEX_DIR, current_first_char));
        let file_skip_list = FileSkip::build_skip_list(skip_list_path);
        FileSkip::write_skip_list(&file_skip_list);
    }
}

File: python/src/lazy_merger.py
from io import TextIOWrapper
import os
from typing import List
from posting import Postings
import json

def get_file_for_word(word: str)->str:
    """Chooses the file to write the word to based on the first character of the word."""
    first_char = word[0].lower()
    if first_char.isalpha():
        return os.path.join(OUTPUT_DIR, f"{first_char}.txt")
    elif first_char.isdigit():
        return os.path.join(OUTPUT_DIR, "0-9.txt")
    # should never have a word starting with a special character or space
    
def list_of_needed_files(batches: List[str]) -> List[TextIOWrapper]:
    # * /indexes/batch_{number}.txt
    return [open(os.path.join(BASEPATH, batch), 'r', encoding='utf-8')
            for batch in batches]


def get_smallest_key(postings: List[Postings]) -> str:
    """Get the smallest key from a list of postings."""
    smallest = postings[0].word
    for posting in postings:
        if posting.word < smallest:
            smallest = posting.word
    return smallest


BASEPATH = "./indexes"
FINAL_INDEX = "final_index.txt"
OUTPUT_DIR = "./merged_indexes"
PHONEBOOK_PATH = "phonebook.json"


if __name__ == "__main__":
    # * Create the output directory if it doesn't exist
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    # go through all the files in indexes folder and merge them into one
    batches = []
    # * /indexes/batch_{number}.txt
    for root, _, files in os.walk(BASEPATH):
        batches.extend(files)
    # now make a reader object but dont read the file yet
    readers = list_of_needed_files(batches)
    file_handlers = {}
    # final_file_appender = open(FINAL_INDEX, 'a+', encoding='utf-8')
    
    # from each reader take a line, see the first token(word),
    # compare it to all others, merge the smaller one first,
    # move to the next line and compare and repeat till all files have
    # been read
    lines = [reader.readline() for reader in readers]
    active_indices = list(range(len(lines)))

    while active_indices:
        # Filter out empty lines (end of file)
        current_indices = [i for i in active_indices if lines[i].strip()]
        # if [] we reached end
        if not current_indices:
            break

        # get all valid postings
        postings = [Postings.construct_postings(
            lines[i]) for i in current_indices]
        # get the smallest key
        smallest = get_smallest_key(postings)
        new_posting = Postings(smallest)

        # Track which readers need to move to next line
        to_advance = []
        for idx in current_indices:
            posting = Postings.construct_postings(lines[idx])
            if posting.word == smallest:
                new_posting = new_posting.merge_postings(posting)
                to_advance.append(idx)

        # Get the output file for the smallest word
        output_file = get_file_for_word(smallest)
        # * Write the merged posting to the appropriate file
        if output_file not in file_handlers:
            file_handlers[output_file] = open(output_file, 'a+', encoding='utf-8')
        
        file_handlers[output_file].write(str(new_posting) + "\n")
        
        # Move to next line for relevant readers
        for idx in to_advance:
            lines[idx] = readers[idx].readline()
            if not lines[idx].strip():  # Reader reached end
                readers[idx].close()
                active_indices.remove(idx)
    
    # Close all file handlers
    for handler in file_handlers.values():
        handler.close()
    
    # also making the phone book of getting the docid to file location
    current_path = "../developer/DEV"
    batches = []
    for root, _, files in os.walk(current_path):
        batches.extend([os.path.join(root, file) for file in files])

    # Create mapping of docid to file location
    docid_to_file = {}
    for i, file_path in enumerate(sorted(batches), 1):
        docid_to_file[i] = file_path

    # Write the mapping to phonebook.json
    with open('phonebook.json', 'w', encoding='utf-8') as f:
        json.dump(docid_to_file, f, indent=2)
